### DISCUSSION TITLE
New File Version after JP is restarted

### DISCUSSION THREAD

**Discussion by Danieliusss** [koechlm/Vault-Job-Processor---iLogic-Extension#35] [Q&A]
Dear Marcus,

I hope this message finds you well. I have been utilizing the JobProcessor iLogic extension for several years, transitioning from the 2022 to the 2024 version recently. I greatly appreciate its capabilities and performance. 

I have encountered a recurring issue:
The JobProcessor is deployed on a separate virtual machine, where it monitors the Job Queue every minute to execute automation tasks triggered by part lifecycle changes (~1000 tasks a day). I have observed that the Job Processor sometimes checks in a new file version after executing an JP iLogic rule. This behavior occurs after restarting the virtual machine (Or JP application) or during periods when no manual iLogic tasks have been submitted (with only lifecycle-triggered tasks running). Notably, this issue has been consistent across both the 2022 and 2024 versions of the Vault plugins.

<img width="664" alt="image" src="https://github.com/koechlm/Vault-Job-Processor---iLogic-Extension/assets/48447888/de048869-8d74-41d7-a00b-cfb23f204f17">

Temporary Workaround:
The only effective workaround currently involves manually submitting an iLogic task on at least one file. This manual intervention seems to correct the behavior for several days, or until the Job Processor application is restarted, during which no unnecessary versions are checked into the Vault post-iLogic task execution.

I would greatly appreciate your guidance on how to permanently resolve this issue or information regarding any available updates that address this behavior. Ensuring that the JobProcessor does not inadvertently check in new file versions after rule execution is crucial for maintaining our workflow efficiency.

Best regards,
Danielius

**Comment by koechlm** [comment]
Hi Danielius, Thank you for the feedback! I am happy to see that the sample and open-source code are valuable. I hope you run your compilation instead of using the provided setup, which is not allowed for production usage.
The lifecycle-triggered jobs create new versions on purpose, so what happens in your environment if a file is dirtied by the job?

**Reply by Danieliusss** [reply]
Hi Marcus,

Thank you for your response. We are running our compilation of the program.

Our Lifecycle Rules/iLogics are designed for tasks like taking HD screenshots, reading iProperties, saving BOMs, etc. and do not alter files. Yet, their execution triggers new file versions, causing these files to appear outdated in Inventor and requiring authors to refresh them.
Additionally, we have an issue with part thumbnails in Vault. Despite using a default white background for all parts on manual Check-in, the thumbnails revert to a blue background when checked in by iLogic, disrupting our documentation which relies on these thumbnails.
Furthermore, when an engineer is using Inventor 2022 and a file is checked in and migrated to 2024 by Vault due to a new version being created, the engineer with Inventor 2022 can no longer open the migrated file. 

Interestingly, running a manuall JP iLogic (User Rule) task temporarily prevents the creation of new file versions even by lifecycle rules until the next Job Processor restart. This observation suggests a potential workaround but is not a sustainable solution.

I understand now that the lifecycle-triggered jobs are designed to create new versions on purpose (not sure why though). However, is there a way to disable new version creation for all unaltered files, both for user rules and lifecycle rules? 

Best regards,
Danielius

**Comment by koechlm** [comment]
Hi Danielius, I agree. You have good reasons to undo check-out on your jobs. There is no reason not to implement this as an option for rule jobs triggered by lifecycles. However, I don't have this on my list for now. The other strange thing that a user rule would prevent a subsequent job triggered by lifecycle might be something to reproduce and investigate. Do you have a secured step-by-step workflow to reproduce it?

**Reply by Danieliusss** [reply]
Hi Marcus,

I have all the mentioned functions in a single iLogic code, which runs every time a lifecycle state is changed as a primary rule. Depending on whether the state is obsolete, released, or any other, different parts of my code in the primary iLogic code are executed. To avoid new file versions being created by lifecycle-activated rules, I use a workaround: running this exact iLogic code as a user rule once on any .ipt file. This method consistently prevents new versions from being created for all files until the Job Processor (JP) is restarted. My JP setup runs on a dedicated virtual machine, which checks for pending tasks every minute and does nothing else. The virtual machine has Inventor 2024 and Vault 2024 installed.

Best regards,
Danielius
